### ISSUE TITLE
mdDialog a11y: confirm and custom dialog main content never gets focus

### ISSUE THREAD

**Issue by reedlabotz** [angular/material#3851]
The main content of the mdDialog doesn't ever get focus.

To reproduce:
* Open https://material.angularjs.org/latest/#/demo/material.components.dialog
* Click CONFIRM DIALOG
* Focus is on PLEASE DO IT
* Pressing tab takes you to browser URL bar
* Pressing tab reads the dialog description
* Pressing tab reads SOUNDS LIKE A SCAM
* Pressing tab takes you back to PLEASE DO IT

I would expect the main content to be the first thing to get focus when the dialog opens.

**Comment by ThomasBurleson** [comment]
@reedlabotz - That demo uses an **confirm** dialog instead of an **alert** dialog.

```
var confirm = $mdDialog.confirm()
          .title('Would you like to delete your debt?')
          .content('All of the banks have agreed to forgive you your debts.')
          .ariaLabel('Lucky day')
          .ok('Please do it!')
          .cancel('Sounds like a scam')
          .targetEvent(ev);

$mdDialog.show(confirm);
```

*  if the dialog type is an alert then the initial focus is on the content. 
*  for all non-alerts, the focus is on the close button.

**Comment by reedlabotz** [comment]
Yeah, looks like you are correct.
I think there is an issue in the flow I described. I believe that the dialog should not ever loose focus, that is to say hitting tab should not take me to the url bar, it should wrap around to the top of the modal.
http://dev.vast.com/jquery-popup-overlay/ is an example of focus staying in the modal.

**Comment by ThomasBurleson** [comment]
@reedlabotz - I agree that the tab flow seems off. Tabbing that includes the URL bar seems very odd. I think perhaps the `tabIndex` attributes are not being set properly.

**Comment by marcysutton** [comment]
Tabindex isn't the issue here, it's focus being programmatically sent to the confirm button. It should probably change to the dialog wrapper, but there is no recommended Material Design focus style for wrapping elements. 

Focus is being trapped in the modal as far as the DOM is concerned, but possibly more work could be done to avoid it going to the address bar.

**Comment by marcysutton** [comment]
I dug up some more information about whether dialogs should be completely trapped to what's in the DOM. Native JS alertdialogs do prevent tabbing to the browser controls, but the only implemented HTML5 `<dialog>` does allow it (according to [Steve Faulkner](https://gitter.im/w3c/a11ySlackers?at=55ac4c0b0b9672bf3694b4d6)). 

When a user has a lot of tabs, it's easy to get lost when tabbing is allowed outside of the browser viewport. But on the other hand, you're preventing them from leaving the window, and doing things in other tabs intentionally, until they have interacted with the dialog. What if you need to retrieve something in another tab before you can confirm or cancel the dialog? Sounds like a good candidate for actual user testing. 

Either way, this project should definitely plan to move to the HTML5 `<dialog>` at some point–it wasn't solid enough when I originally worked on this. How it will be implemented in other browsers remains to be seen, but it's smart to leverage the work being done.
